lua/diffview/vcs/adapters/hg/init.lua

```diff
@@ -1,18 +1,1258 @@
+
+  ---@type LogOptions
+  local log_options = config.get_log_options(
+    single_file,
+    single_file and log_opt.single_file or log_opt.multi_file,
+    "hg"
+  )
+
+  local state = {
+    thread = thread,
+    adapter = self,
+    path_args = log_opt.single_file.path_args,
+    log_options = log_options,
+    prepared_log_opts = prepare_fh_options(self, log_options, single_file),
+    opt = opt,
+    callback = callback,
+    entries = entries,
+    single_file = single_file,
+    resume_lock = false,
+  }
+
+  local function data_callback(status, d, msg)
+    if status == JobStatus.PROGRESS then
+      data[#data+1] = d
+    end
+
+    last_status = status
+    if msg then
+      err_msg = msg
+    end
+    if not state.resume_lock and coroutine.status(thread) == "suspended" then
+      self:handle_co(thread, coroutine.resume(thread))
+    end
+
+    if co_state.shutdown then
+      return true
+    end
+  end
+
+  incremental_fh_data(state, data_callback)
+
+  while true do
+    if not vim.tbl_contains({ JobStatus.SUCCESS, JobStatus.ERROR, JobStatus.KILLED }, last_status)
+        and not data[data_idx] then
+      coroutine.yield()
+    end
+
+    if last_status == JobStatus.KILLED then
+      logger.warn("File history processing was killed.")
+      return
+    elseif last_status == JobStatus.ERROR then
+      callback(entries, JobStatus.ERROR, err_msg)
+      return
+    elseif last_status == JobStatus.SUCCESS and data_idx > #data then
+      break
+    end
+
+    state.cur = data[data_idx]
+
+    state.commit = Commit({
+      hash = state.cur.right_hash,
+      author = state.cur.author,
+      time = tonumber(state.cur.time),
+      time_offset = state.cur.time_offset,
+      rel_date = state.cur.rel_date,
+      ref_names = state.cur.ref_names,
+      subject = state.cur.subject,
+    })
+
+    local ok, status = parse_fh_data(state)
+
+    if not ok then
+      if status == JobStatus.FATAL then
+        return
+      end
+      break
+    end
+
+    data_idx = data_idx + 1
+  end
+
+  callback(entries, JobStatus.SUCCESS)
+end
+
+function HgAdapter:diffview_options(args)
+  local default_args = config.get_config().default_args.DiffviewOpen
+  local argo = arg_parser.parse(vim.tbl_flatten({ default_args, args }))
+  local rev_args = argo.args[1]
+
+  local left, right = self:parse_revs(rev_args, {})
+  if not (left and right) then
+    return
+  end
+
+  local options = {
+    show_untracked = true, -- TODO: extract from hg config
+    selected_file = argo:get_flag("selected-file", { no_empty = true, expand = true })
+      or (vim.bo.buftype == "" and pl:vim_expand("%:p"))
+      or nil,
+  }
+
+  return {left = left, right = right, options = options}
+end
+
+function HgAdapter:rev_to_pretty_string(left, right)
+  if left.type == RevType.CUSTOM then
+    return left.commit
+  elseif right and right.commit then
+    return left:abbrev() .. "::" .. right:abbrev()
+  end
+  return nil
+end
```

Comment:
Did you mean to use `RevType.CUSTOM` here? If so, why? This rev type is a remnant of `:h diffview-api` that I don't think anyone has ever used, and I'm probably going to remove it in the future.

Also, why is the logic in this method different from the logic in the super method?

Reply:
right, it's a leftover from some trials